{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 1: Controls¶\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx). (hint: This will help you decide how to load the data into the database)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'_sre.SRE_Match' object is not callable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-10-ab222907eb9a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0mmatchObj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mre\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msearch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mr'(AT.G.{6})'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mmatchObj\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m         \u001b[0mname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmatchObj\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m         \u001b[0;32mcontinue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: '_sre.SRE_Match' object is not callable"
     ]
    }
   ],
   "source": [
    "germplasmread = open ('Germplasm.tsv', 'r')\n",
    "genelocus = open ('LocusGene.tsv', 'r')\n",
    "\n",
    "import re\n",
    "\n",
    "for line in germplasmread.readlines():\n",
    "    line = line.strip()\n",
    "    matchObj = re.search(r'(AT.G.{6})', line)\n",
    "    if matchObj:\n",
    "        name = matchObj(1)\n",
    "    else:\n",
    "        continue\n",
    "\n",
    "germplasmread.close()\n",
    "\n",
    "for gene in genelocus.readlines():\n",
    "    gene = gene.strip()\n",
    "    matchObj = re.search(r'(AT.G.{6})', gene)\n",
    "    if matchObj:\n",
    "        name2 = matchObj(1)\n",
    "    else:\n",
    "        continue\n",
    "\n",
    "genelocus.close()\n",
    "\n",
    "print(name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 2: Design and create the database.\n",
    "\n",
    "It should have two tables - one for each of the two data files.\n",
    "The two tables should be linked in a 1:1 relationship\n",
    "you may use either sqlMagic or pymysql to build the database\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Locus\\tGene\\tProteinLength\\n', 'AT1G01040\\tDCL1\\t332\\n', 'AT1G01060\\tLHY\\t290\\n', 'AT1G01140\\tCIPK9\\t223\\n', 'AT1G01220\\tFKGP\\t190\\n', 'AT2G03720\\tMRH6\\t189\\n', 'AT2G03800\\tGEK1\\t196\\n', 'AT2G04240\\tXERICO\\t256\\n', 'AT2G05210\\tPOT1A\\t221\\n', 'AT3G02130\\tRPK2\\t284\\n', 'AT3G02140\\tTMAC2\\t300\\n', 'AT3G02230\\tRGP1\\t301\\n', 'AT3G02260\\tBIG\\t279\\n', 'AT3G02310\\tSEP2\\t175\\n', 'AT3G02680\\tNBS1\\t190\\n', 'AT3G02850\\tSKOR\\t234\\n', 'AT3G02870\\tVTC4\\t311\\n', 'AT3G03260\\tHDG8\\t194\\n', 'AT4G14790\\tSUV3\\t312\\n', 'AT4G15210\\tBAM5\\t313\\n', 'AT4G15560\\tDXS\\t219\\n', 'AT4G15570\\tMAA3\\t294\\n', 'AT4G15802\\tHSBP\\t254\\n', 'AT4G15880\\tESD4\\t265\\n', 'AT4G16420\\tADA2B\\t279\\n', 'AT4G16480\\tINT4\\t284\\n', 'AT5G10480\\tPAS2\\t301\\n', 'AT5G10510\\tAIL6\\t310\\n', 'AT5G11110\\tSPS2\\t232\\n', 'AT5G11260\\tHY5\\t221\\n', 'AT5G11510\\tMYB3R-4\\t336\\n', 'AT5G12200\\tPYD2\\t310\\n', 'AT5G13290\\tCRN\\t189\\n']\n"
     ]
    }
   ],
   "source": [
    "#germplasmread = open (\"Germplasm.tsv\", \"r\")\n",
    "#germplasmread.seek(0)\n",
    "#print(germplasmread.readlines())\n",
    "\n",
    "#genelocus = open ('LocusGene.tsv', 'r')\n",
    "#genelocus.seek(0)\n",
    "#print(genelocus.readlines())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "too many values to unpack (expected 3)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-36-487f033ae74d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     63\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     64\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mgenes\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 65\u001b[0;31m                 \u001b[0mlocus\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgene\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mproteinlenght\u001b[0m  \u001b[0;34m=\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     66\u001b[0m                 \u001b[0msql\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"INSERT INTO locus (Locus, Gene, ProteinLenght) VALUES ('%s', '%s', '%d')\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mlocus\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgene\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mproteinlenght\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     67\u001b[0m                 \u001b[0mcursor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msql\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: too many values to unpack (expected 3)"
     ]
    }
   ],
   "source": [
    "#%sql create database germplasmexam;\n",
    "#%sql show databases\n",
    "\n",
    "#%sql use germplasmexam;\n",
    "#%sql show tables\n",
    "\n",
    "#%sql drop table germplasm\n",
    "#%sql CREATE TABLE germplasm(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(10) NOT NULL, Germplasm VARCHAR(10) NOT NULL, Phenotype VARCHAR(100) NOT NULL, Pubmed INTEGER NOT NULL);\n",
    "#%sql DESCRIBE germplasm \n",
    "\n",
    "#%sql SELECT * FROM germplasm\n",
    "#%sql drop table locus\n",
    "#%sql CREATE TABLE locus (id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(10) NOT NULL, Gene VARCHAR(10) NOT NULL, ProteinLenght INTEGER NOT NULL);\n",
    "#%sql show tables\n",
    "import pymysql.cursors\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='germplasmexam',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor,\n",
    "                             autocommit=True) #HAY QUE AÑADIRLO\n",
    "\n",
    "germplasmread = open (\"Germplasm.tsv\", \"r\")\n",
    "germplasmread.seek(0)\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        for germ in germplasmread.readlines():\n",
    "            germ = germ.strip()  #  .strip()  removes spaces and newline at both beginning and end\n",
    "\n",
    "            if germ:\n",
    "                locus, germplasm, phenotype, pubmed = line.split('\\t')\n",
    "                sql = \"INSERT INTO germplasm (Locus, Germplasm, Phenotype, Pubmed) VALUES ('%s', '%s', '%s', '%d')\" % (locus, germplasm, phenotype, int(pubmed))\n",
    "                cursor.execute(sql)\n",
    "            else:\n",
    "                continue  \n",
    "                \n",
    "        germplasmread.close()\n",
    "\n",
    "        sql = \"select * from germplasm\"   \n",
    "        \n",
    "finally:\n",
    "    connection.close()\n",
    "        \n",
    "\n",
    "            \n",
    "    #%sql INSERT INTO germplasm (Locus, Germplasm, Phenotype, Pubmed) VALUES (locus, gene, protein, lenght)\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='germplasmexam',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor,\n",
    "                             autocommit=True) #HAY QUE AÑADIRLO\n",
    "\n",
    "genelocus = open ('LocusGene.tsv', 'r')\n",
    "genelocus.seek(0)\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        for genes in genelocus.readlines():\n",
    "            genes = genes.strip()  #  .strip()  removes spaces and newline at both beginning and end\n",
    "\n",
    "            if genes:\n",
    "                locus, gene, proteinlenght  = line.split('\\t')\n",
    "                sql = \"INSERT INTO locus (Locus, Gene, ProteinLenght) VALUES ('%s', '%s', '%d')\" % (locus, gene, int(proteinlenght))                                                                               \n",
    "                cursor.execute(sql)\n",
    "            else:\n",
    "                continue  \n",
    "                \n",
    "        genelocus.close()\n",
    "\n",
    "        sql = \"select * from locus\"   \n",
    "        \n",
    "finally:\n",
    "    connection.close()\n",
    "\n",
    "#Show all the data in the same table    \n",
    "\n",
    "%sql SELECT * FROM germplasm, locus WHERE germplasm.Locus = locus.Locus;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 3: Fill the database\n",
    "\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database. There are a variety of strategies to accomplish this. I will give all strategies equal credit - do whichever one you are most confident with.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='germplasmexam',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor,\n",
    "                             autocommit=True) #HAY QUE AÑADIRLO\n",
    "\n",
    "germplasmread = open (\"Germplasm.tsv\", \"r\")\n",
    "germplasmread.seek(0)\n",
    "genelocus = open ('LocusGene.tsv', 'r')\n",
    "genelocus.seek(0)\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        for friend in friendfile.readlines():\n",
    "            friend = friend.strip()  #  .strip()  removes spaces and newline at both beginning and end\n",
    "\n",
    "            if friend:\n",
    "                name, age = friend.split(\"\\t\")\n",
    "                print(name + \"and\" + age)\n",
    "                sql = \"INSERT INTO names (name, age) VALUES ('%s', %s)\" % (name, age)\n",
    "                cursor.execute(sql)\n",
    "            else:\n",
    "                continue        \n",
    "\n",
    "                \n",
    "        friendfile.close()\n",
    "\n",
    "        sql = \"select * from names\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall()\n",
    "        for result in results:\n",
    "            print(\"friend: %s  has %s years\" % (result['name'], str(result['age'])))\n",
    "\n",
    "finally:\n",
    "    connection.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 4: Create reports, written to a file\n",
    "\n",
    "    Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "    Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "    Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "    Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule!\n",
    "\n",
    "All reports should be written to the same file. You may name the file anything you wish.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "%sql SELECT * FROM germplasm, locus WHERE germplasm.Locus = locus.Locus;\n",
    "\n",
    " #Create a joined report that only includes the Genes SKOR and MAA3\n",
    "    \n",
    "#Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='germplasmexam',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor,\n",
    "                             autocommit=True) #HAY QUE AÑADIRLO\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        import re\n",
    "sql = \"select * from locus\"\n",
    "cursor.execute(sql)\n",
    "results = cursor.fetchall()\n",
    "matchObj1 = re.search(r'(AT1G)', results)\n",
    "matchObj2 = re.search(r'(AT5G)', results)\n",
    "name = matchObj1(1)\n",
    "if matchObj1(1):\n",
    "    sql = select name avg name AS \"entries for Chromosome 1\"\n",
    "name2 = matchObj2(1)\n",
    "if matchObj2(1):\n",
    "    sql = select name avg name AS \"entries for Chromosome 5\"\n",
    "\n",
    "inally:\n",
    "    connection.close()\n",
    "\n",
    "#Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
